### ISSUE TITLE
Allow the introduction of the atomic transaction to be optional.

### ISSUE THREAD

**Issue by ernestoferro** [etianen/django-reversion#624]
As weird as it seems, wrapping a section of code inside of an atomic transaction can modify the behaviour of an application using django-reversion.

In Django, the changes made inside of a transaction are kept for the thread that initiated that transaction. So if you change a row, and re-read that from you'll get the modified value. If anyone outside of the transaction does the same thing, they'll see the previous value (unless you commit the transaction obviously).

Now, if an application using django-reversion happens to assume that auto-commit is True (and this is the default), it could modify a row in the DB and then launch a thread that assumes that this change has happened. But if the change was done inside of a transaction, this thread will not see the change (there's a race condition there, but that's another topic). This could introduce problems in the project using this library.

While all of this can be avoided (and it probably should), I think that the user of django-reversion should make that call. Especially if it's a big code base where this simple change can be seen as risky.

Adding a setting for this would be very useful.

**Comment by etianen** [comment]
I agree that this needs to be a setting. However, using the setting IS a
bad idea. So this is definitely a foot-gun.

**Comment by ernestoferro** [comment]
True, but Django itself doesn't set ATOMIC_REQUESTS for you, even though it could argue that if you don't you are shooting yourself in the foot. Is up to the developer to decide if AUTO_REQUESTS=True is worth it or not (performance, compatibility with legacy code, etc.)
https://docs.djangoproject.com/en/1.10/ref/settings/#std:setting-DATABASE-ATOMIC_REQUESTS

And I'm comparing it with AUTO_REQUESTS because using the django-reversion middleware is pretty much the same thing. And I'm not sure how many people are aware of the fact that using a third-party library is pretty much setting AUTO_REQUESTS to True.

In fact, if the developer sets AUTO_REQUESTS, wouldn't that achieve what you are trying to achieve with that transaction? Which again I think it should be a decision that the developer of the project should set.

BTW, this wouldn't be an issue if the atomic transaction is only applied the DB operations made in django-reversion.

In summary, wrapping the request inside an atomic transaction when using the middleware is the same as setting AUTO_REQUESTS=True to the project, but this is not done by the developer but by a third party library. Maybe without the user noticing. And if Django doesn't set AUTO_REQUESTS by default, I don't think a library should do the equivalent.

**Comment by etianen** [comment]
Django doesn't set ATOMIC_REQUESTS by default. django-reversion doesn't set
RevisionMiddleware by default either. Both are conveniences that have to be
explicitly enabled by the user.

If you don't want to wrap every request in a transaction, don't use
ATOMIC_REQUESTS. Use the transaction.atomic() decorator.

Likewise, if you don't want to wrap every request in a transaction and a
revision, don't use RevisionMiddleware. Use the reversion.create_revision()
decorator.

I support adding a way of disabling the transaction management in
Django-reversion. However, it will always come with a big warning that it
can break data integrity, and it will never be enabled by default.

**Comment by ernestoferro** [comment]
Thanks Dave.

I guess the point I was trying to make is that the use of the django-reversion middleware and data integrity should independent things. And that users that are about data integrity should set ATOMIC_REQUESTS = True (and they should).

Anyway, I think we are on the same page. I'm all for data integrity, I just think that it should be set by the project, not a library.

Regardless of that, I really appreciate the fact that you are willing to introduce this change.

So, any preference for the setting name?

**Comment by etianen** [comment]
Currently, django-reversion does not use any top-level settings, and I'd like to keep it that way. Instead, I'd implement this feature at all levels of revision management:

1. A new `atomic=True` keyword parameter to `reversion.create_revision()`.
2. A new `atomic=True` keyword parameter to `reversion.views.create_revision()`.
3. A new `revision_atomic=True` class property on `reversion.views.RevisionMixin`.
4. A new `atomic=True` class property on `reversion.middleware.RevisionMiddleware`.

That way, the feature is uniform with all other reversion features, and controlled in the same way.

**Comment by ernestoferro** [comment]
Sounds good. I'll create a fork and send you a pull request and we can iterate on that.
BTW, for number 3, did you mean atomic=True or revision_atomic=True?

**Comment by etianen** [comment]
I meant `revision_atomic=True`.

For 1, 2, and 4, the keywords are on an object or function that is owned by django-reversion, so I don't namespace them.

For 3, because it's a view mixin that will be combined with other views, all reversion-specific keywords are namespaced with `revision_` to prevent conflicts.

I look forward to you pull request. :)

**Comment by ernestoferro** [comment]
Excellent point. I didn't stop to think about it.
Thanks.

**Comment by lampslave** [comment]
Can RevisionMiddleware take into consideration `non_atomic_requests`? Maybe we already have enough ways to disable transactions (decorator and `AUTO_REQUESTS=False`)?

**Comment by ernestoferro** [comment]
Just to clarify, the revision middleware starts the revision before all of
the views (but maybe not of all the other middlewares) start executing. And
it closes the revision in the response after of all of the views have
finished (but again, some other middlewares have execute more things).
Starting a revision also starts an atomic transaction. So in a way, it
behaves almost like ATOMIC_REQUEST=True. And I say almost because there
could be other "outer" middlewares executing before and after revision.

In summary, there are not the same thing. They just happen to have almost
the same outcome. To the point where setting atomic transaction to true
will make no difference if the revisions middleware is in use.

I hope this clarifies the situation.

**Comment by lampslave** [comment]
Thank you for the clarification. I knew that middleware "wraps" view, but was hoping for any trick.

**Comment by etianen** [comment]
Closed by #625

**Comment by AndersonDan** [comment]
How to set atomic=False in middleware?

**Comment by ernestoferro** [comment]
You have either change the value directly in the class `RevisionMiddleware.atomic = True`
or subclass it and change it there.

**Comment by AndersonDan** [comment]
I figured this was it, I just thought there was something in django's "MIDDLEWARE" settings that was not documented. Thanks for the answer.
